File: tetris_compose/app/src/main/java/com/robin/bachelor/thesis/tetris_compose/Tetromino.kt
package com.robin.bachelor.thesis.tetris_compose

import kotlin.random.Random

class Tetromino {
  private var position = Pair(GameLogic.GRID_WIDTH / 2, 0)
  private var shape = Random.nextInt(3) // 0 = square, 1 = L-shape, 2 = T-shape
  private var rotate = 0

  fun getBlocks(dx: Int, dy: Int, dr: Int): List<Pair<Int, Int>> {
    val x = position.first + dx
    val y = position.second + dy
    val r = (rotate + dr) % 4

    return when (shape) {
      0 -> listOf(Pair(x, y), Pair(x + 1, y), Pair(x, y + 1), Pair(x + 1, y + 1)) // Square
      1 -> when (r) {
        0 -> listOf(Pair(x, y), Pair(x + 1, y), Pair(x - 1, y), Pair(x - 1, y + 1))
        1 -> listOf(Pair(x, y), Pair(x, y + 1), Pair(x, y - 1), Pair(x - 1, y - 1))
        2 -> listOf(Pair(x, y), Pair(x - 1, y), Pair(x + 1, y), Pair(x + 1, y - 1))
        else -> listOf(Pair(x, y), Pair(x, y - 1), Pair(x, y + 1), Pair(x + 1, y + 1))
      }
      else -> when (r) { // T-Shape
        0 -> listOf(Pair(x, y), Pair(x - 1, y), Pair(x + 1, y), Pair(x, y + 1))
        1 -> listOf(Pair(x, y), Pair(x, y - 1), Pair(x, y + 1), Pair(x - 1, y))
        2 -> listOf(Pair(x, y), Pair(x - 1, y), Pair(x + 1, y), Pair(x, y - 1))
        else -> listOf(Pair(x, y), Pair(x, y - 1), Pair(x, y + 1), Pair(x + 1, y))
      }
    }
  }

  fun moveLeft() { position = Pair(position.first - 1, position.second) }

  fun moveRight() { position = Pair(position.first + 1, position.second) }

  fun moveDown() { position = Pair(position.first, position.second + 1) }

  fun rotate() { rotate = (rotate + 1) % 4 }
}

File: tetris_flutter/lib/tetromino.dart
import 'dart:math';

class Tetromino {
  Point<int> position;
  int shape; // 0 = square, 1 = L-shape, 2 = T-shape
  int rotate; // 0, 1, 2, 3

  Tetromino({required this.position, required this.shape, required this.rotate});

  List<Point<int>> getBlocks(int dx, int dy, int dr) {
    int positionXNew = position.x + dx;
    int positionYNew = position.y + dy;
    int rotateNew = (rotate + dr) % 4;

    switch (shape) {
      case 0: // Square
        return [
          Point(positionXNew, positionYNew),
          Point(positionXNew + 1, positionYNew),
          Point(positionXNew, positionYNew + 1),
          Point(positionXNew + 1, positionYNew + 1)
        ];
      case 1: { // L-Shape
        switch(rotateNew) {
          case 0:
            return [
              Point(positionXNew, positionYNew),
              Point(positionXNew + 1, positionYNew),
              Point(positionXNew - 1, positionYNew),
              Point(positionXNew - 1, positionYNew + 1)
            ];
          case 1:
            return [
              Point(positionXNew, positionYNew),
              Point(positionXNew, positionYNew + 1),
              Point(positionXNew, positionYNew - 1),
              Point(positionXNew - 1, positionYNew -1)
            ];
          case 2:
            return [
              Point(positionXNew, positionYNew),
              Point(positionXNew - 1, positionYNew),
              Point(positionXNew + 1, positionYNew),
              Point(positionXNew + 1, positionYNew - 1)
            ];
          case 3:
            return [
              Point(positionXNew, positionYNew),
              Point(positionXNew, positionYNew - 1),
              Point(positionXNew, positionYNew + 1),
              Point(positionXNew + 1, positionYNew + 1)
            ];
          default:
            return [];
        }
      }
      case 2: { // T-Shape
        switch(rotateNew) {
          case 0:
            return [
              Point(positionXNew, positionYNew),
              Point(positionXNew - 1, positionYNew),
              Point(positionXNew + 1, positionYNew),
              Point(positionXNew, positionYNew + 1)
            ];
          case 1:
            return [
              Point(positionXNew, positionYNew),
              Point(positionXNew, positionYNew - 1),
              Point(positionXNew, positionYNew + 1),
              Point(positionXNew - 1, positionYNew)
            ];
          case 2:
            return [
              Point(positionXNew, positionYNew),
              Point(positionXNew - 1, positionYNew),
              Point(positionXNew + 1, positionYNew),
              Point(positionXNew, positionYNew - 1)
            ];
          case 3:
            return [
              Point(positionXNew, positionYNew),
              Point(positionXNew, positionYNew - 1),
              Point(positionXNew, positionYNew + 1),
              Point(positionXNew + 1, positionYNew)
            ];
          default:
            return [];
        }
      }
      default:
        return [];
    }
  }

  void moveDown() {
    position = Point(position.x, position.y + 1);
  }

  void moveLeft() {
    position = Point(position.x - 1, position.y);
  }

  void moveRight() {
    position = Point(position.x + 1, position.y);
  }

  void rotateClockwise() {
    rotate = (rotate + 1) % 4;
  }
}